src/main/java/com/github/retro_game/retro_game/service/impl/ResetPasswordServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+import com.github.retro_game.retro_game.entity.UserPasswordResetTokenKey;
+import com.github.retro_game.retro_game.repository.UserPasswordResetTokenRepository;
+import com.github.retro_game.retro_game.repository.UserRepository;
+import com.github.retro_game.retro_game.service.EmailService;
+import com.github.retro_game.retro_game.service.ResetPasswordService;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.commons.lang3.time.DateUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Date;
+import java.util.Optional;
+
+@Service
+public class ResetPasswordServiceImpl implements ResetPasswordService {
+  private final PasswordEncoder passwordEncoder;
+  private final UserRepository userRepository;
+  private final EmailService emailService;
+  private final UserPasswordResetTokenRepository userPasswordResetTokenRepository;
+  private final int passwordResetLinkExpiresAfterDays;
+
+  public ResetPasswordServiceImpl(
+      @Value("${retro-game.password-reset-link-expires-after-days}") int passwordResetLinkExpiresAfterDays,
+      PasswordEncoder passwordEncoder,
+      UserRepository userRepository,
+      EmailService emailService,
+      UserPasswordResetTokenRepository userPasswordResetTokenRepository) {
+    this.passwordEncoder = passwordEncoder;
+    this.userRepository = userRepository;
+    this.emailService = emailService;
+    this.userPasswordResetTokenRepository = userPasswordResetTokenRepository;
+    this.passwordResetLinkExpiresAfterDays = passwordResetLinkExpiresAfterDays;
+  }
+
+  @Override
+  @Transactional
+  public void generateTokenAndSendEmail(String email) {
+    Optional<User> user = userRepository.findByEmailIgnoreCase(email);
+
+    user.ifPresent(u -> {
+      String plainToken = generateToken();
+      Optional<UserPasswordResetToken> existingToken = userPasswordResetTokenRepository.findByKey_User(u);
+
+      if (existingToken.isPresent()) {
+        userPasswordResetTokenRepository.save(
+            updatedToken(plainToken, existingToken.get())
+        );
+      } else {
+        userPasswordResetTokenRepository.save(
+            prepareNewToken(u, plainToken)
+        );
+      }
+
+      emailService.sendResetPasswordEmail(u, plainToken);
+    });
+  }
+
+  @Override
+  @Transactional
+  public void resetUserPassword(String plainToken, String newPassword) {
+    Optional<UserPasswordResetToken> token = userPasswordResetTokenRepository.findAll().stream()
+        .filter(t -> passwordEncoder.matches(plainToken, t.getEncryptedToken()))
+        .findFirst();
+
+    token.ifPresent(t -> {
+      if (!isTokenExpired(t)) {
+        updatePassword(newPassword, t);
+      } else {
+        userPasswordResetTokenRepository.delete(t);
+      }
+    });
+
```

Comment:
It seems that both `generateTokenAndSendEmail()` and `resetUserPassword()` will fail silently if the email was not sent or the token is wrong/expired, and the user will get a successful message. We could add some error handling, but this could be done in another PR. Could you open an issue for that after this is merged so that we don't forget about it? Or let me know and I will open one.

Reply:
Generally speaking, its a feature :D

In case of `generateTokenAndSendEmail` I wanted to hide information about email presence in DB. In example, if someone will want to check if some email is present in game, he will always get successfull message, even if there is no user with this email.

Same for `resetUserPassword` - if token is incorrect, user will get success message. It's in case of brute force attacks on tokens collection.

Wondering if we should present some error page when email was not send due to ie. smtp error. Then we should inform user instead of returning blank 500 page. I think that it should be an issue.

Reply:
We shouldn't sacrifice the experience of non-malicious users for security.

> In case of generateTokenAndSendEmail I wanted to hide information about email presence in DB. In example, if someone will want to check if some email is present in game, he will always get successfull message, even if there is no user with this email.

Agreed, if there is no email in our database, we should print: the email was sent to x@x.

> Wondering if we should present some error page when email was not send due to ie. smtp error. Then we should inform user instead of returning blank 500 page. I think that it should be an issue.

We should. If I am not mistaken, right now we don't even return 500, just log the error and say everything is fine.

> Same for resetUserPassword - if token is incorrect, user will get success message. It's in case of brute force attacks on tokens collection.

This is a bad experience for normal users. If the user asks for password reset 2 times, gets 2 emails, and uses the one with wrong token, the user will get 'password was reset' message, while this is not true. Similarly, if the token is expired, the user would think that the password was reset anyway.

I wouldn't care too much about brute forcing the tokens, since each token is now 128-bit, and brute forcing is intractable. Even so, we might add some additional security mechanism later, e.g. captcha after 5 tries.

Reply:
I'll open the issues.